cocotb/decorators.py

```diff
@@ -437,9 +437,9 @@ class test(_py_compat.with_metaclass(_decorator_helper, coroutine)):
             value representing simulation timeout (not implemented).
         expect_fail (bool, optional):
             Don't mark the result as a failure if the test fails.
-        expect_error (bool, optional):
+        expect_error (Union[bool, Exception], optional):
             Don't mark the result as an error if an error is raised.
-            This is for cocotb internal regression use 
+            This is for cocotb internal regression use
             when a simulator error is expected.
```

Comment:
Was just about to merge - but I think this could do with a sentence explaining what passing an exception types or tuple of exceptions means.

Also, strictly this does not take `Union[bool, Exception]` - it takes `Union[bool, Type[Exception], Tuple[Type[Exception, ], ...]]` Perhaps this is better in english after all, as:
```
expect_error (bool or exception type or tuple of exception types, optional):
    If true, consider this test passing if it raises an exception, and failing if it does not.
    If an exception type or tuple of exception types, this gives the specific type of exception that is considered passing.
    This is primarily for cocotb internal regression use when a simulator error is expected.

    For use elsewhere, consider::

        @cocotb.test()
        def my_test(dut):
            try:
                thing_that_should_fail()
            except ExceptionIExpect:
                pass
            else:
                assert False, "Exception did not occur"
```

Reply:
Is there is some particular reason this functionality should be reserved for internal use only? I'm think we should just leave the last part off.

Reply:
I suppose not. What I'd like to encourage is that `expect_fail` is used either for:

* Tests which ought to pass, but are known to fail and haven't been fixed yet (like `@pytest.mark.xfail`
* Tests internal to cocotb that test scheduler and regression manager errors which can't otherwise be caught (such as your timeout-level stuff)

What I don't think this should be used for is testing that user APIs throw exceptions - you can write much better tests using the try / except formulation (ie checking error messages, multiple exceptions checked in a single test, etc.

For that reason, I'd like to try and include an example like the above.